File: src/rust/lib/inventory-ordering-service/inventoryOrderingService.ts
//
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2024 Datadog, Inc.
//

import { ISecret } from "aws-cdk-lib/aws-secretsmanager";
import { Construct } from "constructs";
import { SharedProps } from "../constructs/sharedFunctionProps";
import { InstrumentedLambdaFunction } from "../constructs/lambdaFunction";
import { IFunction } from "aws-cdk-lib/aws-lambda";
import { IEventBus } from "aws-cdk-lib/aws-events";
import {
  SnsEventSource,
} from "aws-cdk-lib/aws-lambda-event-sources";
import { Topic } from "aws-cdk-lib/aws-sns";
import { StringParameter } from "aws-cdk-lib/aws-ssm";
import { LogGroup } from "aws-cdk-lib/aws-logs";
import { RemovalPolicy, Tags } from "aws-cdk-lib";
import {
  DefinitionBody,
  StateMachine,
  LogLevel,
} from "aws-cdk-lib/aws-stepfunctions";

export interface InventoryOrderingServiceProps {
  sharedProps: SharedProps;
  ddApiKeySecret: ISecret;
  sharedEventBus: IEventBus;
}

export class InventoryOrderingService extends Construct {
  inventoryOrderingWorkflowTrigger: IFunction;

  constructor(
    scope: Construct,
    id: string,
    props: InventoryOrderingServiceProps
  ) {
    super(scope, id);

    const newProductAddedTopicParam = StringParameter.fromStringParameterName(
      this,
      "RustNewProductAddedTopicParam",
      "/rust/inventory/product-added-topic"
    );

    const topic = Topic.fromTopicArn(
      this,
      "RustNewProductAddedTopic",
      newProductAddedTopicParam.stringValue
    );

    const workflowLogGroup = new LogGroup(
      this,
      "RustInventoryOrderingServiceLogGroup",
      {
        logGroupName: `/aws/vendedlogs/states/RustInventoryOrderingServiceLogGroup-${props.sharedProps.environment}`,
        removalPolicy: RemovalPolicy.DESTROY,
      }
    );

    const workflow = new StateMachine(this, "RustInventoryOrderingService", {
      stateMachineName: `RustInventoryOrderingService-${props.sharedProps.environment}`,
      definitionBody: DefinitionBody.fromFile(
        "./lib/inventory-ordering-service/workflows/workflow.sample.asl.json"
      ),
      logs: {
        destination: workflowLogGroup,
        includeExecutionData: true,
        level: LogLevel.ALL,
      },
    });
    Tags.of(workflow).add("DD_ENHANCED_METRICS", "true");
    Tags.of(workflow).add("DD_TRACE_ENABLED", "true");

    this.inventoryOrderingWorkflowTrigger = new InstrumentedLambdaFunction(
      this,
      "RustInventoryOrderingWorkflowTrigger",
      {
        sharedProps: props.sharedProps,
        functionName: "RustInventoryOrderingWorkflow",
        handler: "index.handler",
        environment: {
          ORDERING_SERVICE_WORKFLOW_ARN: workflow.stateMachineArn,
        },
        manifestPath: "./src/inventory-ordering/lambdas/product_added_handler/Cargo.toml"
      }
    ).function;
    this.inventoryOrderingWorkflowTrigger.addEventSource(
      new SnsEventSource(topic)
    );
    workflow.grantStartExecution(this.inventoryOrderingWorkflowTrigger);
  }
}

File: src/java/cdk/src/main/java/com/cdk/inventory/ordering/InventoryOrderingService.java
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2024 Datadog, Inc.
 */

package com.cdk.inventory.ordering;

import com.cdk.constructs.InstrumentedFunction;
import com.cdk.constructs.InstrumentedFunctionProps;
import org.jetbrains.annotations.NotNull;
import software.amazon.awscdk.RemovalPolicy;
import software.amazon.awscdk.Tags;
import software.amazon.awscdk.services.lambda.IFunction;
import software.amazon.awscdk.services.lambda.eventsources.SnsEventSource;
import software.amazon.awscdk.services.lambda.eventsources.SnsEventSourceProps;
import software.amazon.awscdk.services.logs.LogGroup;
import software.amazon.awscdk.services.logs.LogGroupProps;
import software.amazon.awscdk.services.sqs.Queue;
import software.amazon.awscdk.services.sqs.QueueProps;
import software.amazon.awscdk.services.ssm.StringParameter;
import software.amazon.awscdk.services.ssm.StringParameterProps;
import software.amazon.awscdk.services.stepfunctions.*;
import software.constructs.Construct;

import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;

public class InventoryOrderingService extends Construct {
    public InventoryOrderingService(@NotNull Construct scope, @NotNull String id, @NotNull InventoryOrderingServiceProps props) {
        super(scope, id);

        LogGroup workflowLogGroup = new LogGroup(this, "JavaInventoryOrderingWorkflowLogGroup", LogGroupProps.builder()
                .logGroupName(String.format("/aws/vendedlogs/states/JavaInventoryOrderingWorkflowLogGroup-%s", props.sharedProps().env()))
                .removalPolicy(RemovalPolicy.DESTROY)
                .build());

        String workflowFilePath = "../cdk/src/main/java/com/cdk/inventory/ordering/workflows/workflow.sample.asl.json";

        StateMachine workflow = new StateMachine(this, "JavaInventoryOrderingWorkflow", StateMachineProps.builder()
                .stateMachineName(String.format("JavaInventoryOrderingService-%s", props.sharedProps().env()))
                .definitionBody(DefinitionBody.fromFile(workflowFilePath))
                .logs(LogOptions.builder()
                        .destination(workflowLogGroup)
                        .includeExecutionData(true)
                        .level(LogLevel.ALL)
                        .build())
                .build());

        Tags.of(workflow).add("DD_ENHANCED_METRICS", "true");
        Tags.of(workflow).add("DD_TRACE_ENABLED", "true");

        HashMap<String, String> functionEnvVars = new HashMap<>(2);
        functionEnvVars.put("DD_SERVICE_MAPPING", String.format("lambda_sns:%s", props.newProductAddedTopic().getTopicName()));
        functionEnvVars.put("ORDERING_SERVICE_WORKFLOW_ARN", workflow.getStateMachineArn());
        
        String compiledJarFilePath = "../inventory-ordering-service/target/com.inventory.ordering-0.0.1-SNAPSHOT-aws.jar";
        IFunction handleProductAddedFunction = new InstrumentedFunction(this, "ProductAddedFunction",
                new InstrumentedFunctionProps(props.sharedProps(), "com.inventory.ordering", compiledJarFilePath, "handleNewProductAdded", functionEnvVars)).getFunction();
        workflow.grantStartExecution(handleProductAddedFunction);
        
        handleProductAddedFunction.addEventSource(new SnsEventSource(props.newProductAddedTopic(), SnsEventSourceProps.builder()
                .deadLetterQueue(new Queue(this, "NewProductAddedEventSourceDLQ", QueueProps.builder()
                        .queueName(String.format("NewProductAddedEventSourceDLQ-%s", props.sharedProps().env()))
                        .build()))
                .build()));
                
        StringParameter workflowArnParam = new StringParameter(this, "InventoryOrderingStateMachineArn", StringParameterProps.builder()
                .parameterName(String.format("/java/%s/inventory-ordering/state-machine-arn", props.sharedProps().env()))
                .stringValue(workflow.getStateMachineArn())
                .build());
        
    }
}
